tarentula/datashare_client.py

```diff
@@ -194,3 +194,25 @@ def temporary_project(self, source=DATASHARE_DEFAULT_PROJECT, delete=True):
             if delete and project is not None:
                 self.delete_index(project)
         return project
+
+    def metadata_fields(self, index=DATASHARE_DEFAULT_PROJECT, document_type='Document'):
+        url = urljoin(self.elasticsearch_host, index)
+        mapping = requests.get(url, cookies=self.cookies, headers=self.headers).json()
+        results = []
+        self.get_fields(document_type, index, mapping, results, [])
+        return results
+
+    def get_fields(self, document_type, index, mapping, results, field_stack):
```

Comment:
We should avoid mutating method parameters and return a copy instead.

Reply:
Just a note here : we did this to make the method [tail recursive](https://stackoverflow.com/questions/33923/what-is-tail-recursion) that means that the recursion is not feeding the stack, and is behaving like a loop : some interpreters are detecting tail recursion and so they can remove the previous stack call for each recursive call.

It is [not applicable for python](https://stackoverflow.com/questions/13591970/does-python-optimize-tail-recursion). Besides this, with the current mapping for datashare we will not have more than 1 depth in the call stack (for tika metadata). So the current implementation is ok.